File: Principal Component Analysis.R
# Principal Component Analysis (PCA)

# Importing the dataset
ds = read.csv('Wine.csv')

# Splitting the dataset into a training set and test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(ds$Customer_Segment, SplitRatio = 0.8)
trs = subset(ds, split == TRUE) # Training Set
tes = subset(ds, split == FALSE) # Test Set

# Feature scaling (if needed)
trs[-14] = scale(trs[-14])
tes[-14] = scale(tes[-14])


# Applying PCA
# install.packages('caret')
# install.packages('PreProcess')
library(caret)
# install.packages('e1071')
library(e1071)

pca = preProcess(x = trs[-14], method = 'pca', pcaComp = 2)
# Training Set
trs = predict(pca, trs)
trs = trs[c(2, 3, 1)]
# Test set
tes = predict(pca, tes)
tes = tes[c(2, 3, 1)]

# Fitting PCA to the training set
# install.packages('e1071')
library(e1071)
cl = svm(formula = CustomerSegmant ~ .,
         data = trs,
         type = 'C-classification',
         kernal = 'linear')

# Predicting the Test set Results
y_pred = predict(cl, newdata = tes[-3])

# Making the Confusion Matrix
cm = table(tes[, 3], y_pred)


# Visualising the Training set results
# install.packages('ElemStatLearn')
library(ElemStatLearn)
set = trs
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('PC1', 'PC2')
y_grid = predict(cl, newdata = grid_set)
plot(set[, -3],
     main = 'Support Vector Machine (Training set)',
     xlab = 'PC1', ylab = 'PC2',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 2, 'deepskyblue', ifelse(y_grid == 1, 'springgreen3', 'tomato')))
points(set, pch = 21, bg = ifelse(set[, 3] == 2, 'blue3', ifelse(set[, 3] == 1, 'cyan', 'purple')))


# Visualising the Test set results
library(ElemStatLearn)
set = tes
X1 = seq(min(set[, 1]) - 1, max(set[, 1]) + 1, by = 0.01)
X2 = seq(min(set[, 2]) - 1, max(set[, 2]) + 1, by = 0.01)
grid_set = expand.grid(X1, X2)
colnames(grid_set) = c('PC1', 'PC2')
y_grid = predict(cl, newdata = grid_set)
plot(set[, -3], main = 'Support Vector Machine (Test set)',
     xlab = 'PC1', ylab = 'PC2',
     xlim = range(X1), ylim = range(X2))
contour(X1, X2, matrix(as.numeric(y_grid), length(X1), length(X2)), add = TRUE)
points(grid_set, pch = '.', col = ifelse(y_grid == 2, 'deepskyblue', ifelse(y_grid == 1, 'springgreen3', 'tomato')))
points(set, pch = 21, bg = ifelse(set[, 3] == 2, 'blue3', ifelse(set[, 3] == 1, 'cyan', 'purple')))

File: Principal Component Analysis.py
# Principal Component Analysis (PCA)


# Import the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
ds = pd.read_csv("Wine.csv")
X = ds.iloc[:, 0:13].values
Y = ds.iloc[:, 13].values

# Splitting the dataset into a training set and test set
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.20, random_state = 0)

# Feature scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# Apply dimensionality reduction (PCA)
from sklearn.decomposition import PCA
pca = PCA(n_components = 2) # Extract the top 2 principle components
X_train = pca.fit_transform(X_train)
X_test = pca.transform(X_test)
# Explained Variance
exv = pca.explained_variance_ratio_


# Fitting the PCA to the Training Set (logistic regression)
from sklearn.linear_model import LogisticRegression
cl = LogisticRegression(random_state = 0)
cl.fit(X_train, Y_train)


# Predicint the test set results
y_pred = cl.predict(X_test)

# Making the confustion matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test, y_pred)

# Visualising the Training set results
from matplotlib.colors import ListedColormap
X_set, Y_set = X_train, Y_train
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, cl.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('darkviolet', 'green', 'goldenrod')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(Y_set)):
    plt.scatter(X_set[Y_set == j, 0], X_set[Y_set == j, 1],
                c = ListedColormap(('pink', 'springgreen', 'gold'))(i), label = j)
plt.title('Principal Component Analysis (Training set)')
plt.xlabel('PC1')
plt.ylabel('PC2')
plt.legend()
plt.show()

# Visualising the Test set results
from matplotlib.colors import ListedColormap
X_set, Y_set = X_test, Y_test
X1, X2 = np.meshgrid(np.arange(start = X_set[:, 0].min() - 1, stop = X_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = X_set[:, 1].min() - 1, stop = X_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, cl.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('darkviolet', 'green', 'goldenrod')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(Y_set)):
    plt.scatter(X_set[Y_set == j, 0], X_set[Y_set == j, 1],
                c = ListedColormap(('pink', 'springgreen', 'gold'))(i), label = j)
plt.title('Principal Component Analysis (Test set)')
plt.xlabel('PC1')
plt.ylabel('PC2')
plt.legend()
plt.show()